### PR TITLE
Select gas price

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import {GasMode, GasOption} from '../models/gasMode';
+
+export const getGasPriceFromGasModes = (gasModes: GasMode[], gasModeName: string, gasTokenAddress: string) => {
+  const mode = gasModes!.filter((gasPriceMode: GasMode) => gasPriceMode.name === gasModeName)[0];
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: `filter` -> `find`

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  const gasOption = mode && mode.gasOptions.filter((gasOption: GasOption) => gasOption.token.address === gasTokenAddress)[0];
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: `mode && mode.` -> `mode!.` should throw on invalid input not silently fail

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  return gasOption && gasOption.gasPrice;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: Unnecessary `gasOption &&` #failfast

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import {GasMode, GasOption} from '../models/gasMode';
+
+export const getGasPriceFromGasModes = (gasModes: GasMode[], gasModeName: string, gasTokenAddress: string) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: `getGasPriceFor()`

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import {expect} from 'chai';
+import {getGasPriceFromGasModes} from '../../../lib/core/utils/getGasPriceFromGasModes';
+import {TEST_GAS_MODES} from '../../../lib/core/constants/test';
+
+describe('getTransactionGasPrice', () => {
+  it('empty array', () => {
+    expect(getGasPriceFromGasModes([], '', '')).to.be.undefined;
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: should throw

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  });
+
+  it('invalid mode name', () => {
+    expect(getGasPriceFromGasModes(TEST_GAS_MODES, 'invalidModeName', TEST_GAS_MODES[0].gasOptions[0].token.address)).to.be.undefined;
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: should throw

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  });
+
+  it('invalid token address', () => {
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: should throw

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 8]
```diff
@@ -99,15 +102,15 @@ export const EmojiForm = ({sdk, contractAddress, privateKey, hideTitle, classNam
...
+          <img className="correct-input-img" src={CheckmarkIcon} alt="checkmark" />
           <p className="correct-input-title">Correct!</p>
           <EmojiPlaceholders
             enteredCode={enteredCode}
             onEmojiClick={onEmojiRemove}
             className={className}
           />
           <div className="correct-input-footer">
-            <GasPrice />
+            <GasPrice sdk={sdk} setGasTokenAddress={setGasToken} setGasPrice={setGasPrice}/>
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]:
```ts
<... onGasTokenChanged="..." ... onGasPriceChanged="..."/>
```

Comment (marekkirejczyk) [reply universal-login-react/src/ui/Notifications/EmojiForm.tsx]: or better:
```ts
<...onGasParametersChange="setGasParametes" />
````

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 9]
```diff
@@ -56,7 +59,7 @@ export const EmojiForm = ({sdk, contractAddress, privateKey, hideTitle, classNam
   };
 
   const confirmCode = (address: string) => {
-    sdk.addKey(contractAddress, address, privateKey, transactionDetails);
+    sdk.addKey(contractAddress, address, privateKey, {...transactionDetails, gasToken, gasPrice});
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]:
```ts
{...transactionDetails, ...gasParamiters}
```

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 10]
```diff
@@ -30,6 +31,8 @@ interface EmojiFormProps {
 }
 
 export const EmojiForm = ({sdk, contractAddress, privateKey, hideTitle, className}: EmojiFormProps) => {
+  const [gasToken, setGasToken] = useState<string>('');
+  const [gasPrice, setGasPrice] = useState<utils.BigNumber>(utils.bigNumberify('0'));
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]:
```ts
type GasParameters {
  gasToken: string,
  gasPrice: utils.BigNumber
}
const EMPTY_GAS_PARAMETERS = {
  gasToken: '0x00.00,
  gasPrice: 0
}
```

```ts
const [gasParameters, setGasParameters] = useState<utils.BigNumber>();
```

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts (hunk 2 above)
Comment (xitronix) [reply universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts (hunk 3 above)
Comment (xitronix) [reply universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts (hunk 4 above)
Comment (xitronix) [reply universal-login-commons/lib/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts (hunk 5 above)
Comment (xitronix) [reply universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts (hunk 6 above)
Comment (xitronix) [reply universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts (hunk 7 above)
Comment (xitronix) [reply universal-login-commons/test/core/utils/getGasPriceFromGasModes.ts]: ✔️

File: universal-login-react/src/ui/commons/GasPrice.tsx [hunk 11]
```diff
@@ -1,21 +1,55 @@
...
+      setModeName(name);
+      onTokenChanged(address, gasPrice);
+    }
+  }, [gasModes]);
   const [contentVisibility, setContentVisibility] = useState(false);
-  const [speedType, setSpeedType] = useState('cheap');
-  const [fee, setFee] = useState('dai');
 
-  return (
+  const formatWeiToEther = (wei: utils.BigNumber) => wei.div('1000000000').toNumber() / 1000000000;
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/commons/GasPrice.tsx]: Use ethers.js function `utils.formatEther`

File: universal-login-react/src/ui/commons/GasPrice.tsx [hunk 12]
```diff
@@ -35,82 +69,61 @@ export const GasPrice = () => {
...
-                    onChange={() => setFee('dai')}
-                  >
-                    <div className="transaction-fee-row">
-                      <div className="transaction-fee-details">
-                        <img src="" alt="" className="transaction-fee-item-icon"/>
-                        <div>
-                          <p className="transaction-fee-amount">0.09 DAI</p>
-                          <p className="transaction-fee-amount-usd">0.09 USD</p>
+
+                {gasModes.filter(gasMode => gasMode.name === modeName)[0].gasOptions.map(({token, gasPrice}) => (
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/commons/GasPrice.tsx]: filter -> find

File: universal-login-sdk/lib/core/services/GasModeService.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import {PriceObserver} from '../observers/PriceObserver';
+
+export class GasModeService {
+  constructor(
+    private tokensStore: TokensDetailsStore,
+    private gasPriceOracle: GasPriceOracle,
+    private priceObserver: PriceObserver
+    ) {}
+
+  private async getMode(name: string, gasPrice: utils.BigNumber, tokensPrices: TokensPrices): Promise<GasMode> {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/services/GasModeService.ts]: unnecessary `async`

Comment (marekkirejczyk) [reply universal-login-sdk/lib/core/services/GasModeService.ts]: `getMode` -> `createMode`

get -> getter
create -> factory method

**Comment by marekkirejczyk** [comment]
I'll merge this.
Please address comments in the following PR.

**Pull request merged**